Add PSBT (BIP-174) support

JAGADISHSUNILPEDNEKAR opened the pull request:
# Add PSBT (BIP-174) Support

## Description
This PR adds support for Partially Signed Bitcoin Transactions (PSBT) as defined in BIP-174. PSBT is a critical format used by hardware wallets and multi-signature setups to coordinate transaction signing across multiple devices or parties.

## Features Added
- `PSBT` class for creating, parsing, and updating PSBTs
- `PSBTInput` and `PSBTOutput` classes for handling PSBT inputs and outputs
- Methods for PSBT serialization and deserialization (bytes, base64, hex)
- Support for adding various input and output types
- Methods for signing inputs
- Support for PSBT finalization and extraction of signed transactions
- Support for combining multiple PSBTs
- Comprehensive test suite for PSBT functionality

## Motivation
Hardware wallet support and multi-signature capability are crucial for secure Bitcoin management. PSBTs enable wallet developers to:
- Coordinate transaction signing across multiple devices
- Support air-gapped signing workflows
- Enable multi-party transaction coordination
- Interface with hardware wallets

## Example Usage
```python
from bitcoinutils.setup import setup
from bitcoinutils.transactions import Transaction, TxInput, TxOutput
from bitcoinutils.keys import PrivateKey
from bitcoinutils.script import Script
from bitcoinutils.psbt import PSBT

# Setup network
setup('testnet')

# Create an unsigned transaction
txin = TxInput('previous_txid', 0)
private_key = PrivateKey('private_key_wif')
address = private_key.get_public_key().get_address()
txout = TxOutput(1000000, address.to_script_pub_key())
tx = Transaction([txin], [txout])

# Create a PSBT from the transaction
psbt = PSBT.from_transaction(tx)

# Add UTXO information
previous_tx = Transaction.from_hex('previous_tx_hex')
psbt.add_input_utxo(0, previous_tx)

# Sign the input
psbt.sign_input(private_key, 0)

# Finalize the PSBT
psbt.finalize()

# Extract the final signed transaction
final_tx = psbt.extract_transaction()

# Get the serialized transaction ready for broadcast
tx_hex = final_tx.serialize()

```

## Testing
All PSBT functionality is covered by unit tests in `test_psbt.py`. Tests include:
- Basic PSBT creation
- PSBT serialization/deserialization
- Input and output operations
- Signing inputs
- Finalizing PSBTs
- Extracting transactions
- Combining PSBTs

## Future Enhancements (Not in this PR)
- Taproot PSBT support (BIP 341/342)
- Integration with Miniscript for more complex spending conditions
- Additional convenience methods for common PSBT operations

## References
- [BIP-174 Specification](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki)

## Screenshots
This shows that the implementation has been successful and all test cases are passing:
<img width="1440" alt="Screenshot 2025-03-08 at 3 14 23 PM" src="https://github.com/user-attachments/assets/96ba518f-be56-4f8e-8f1f-e686741b4b90" />
<img width="1440" alt="Screenshot 2025-03-08 at 3 14 14 PM" src="https://github.com/user-attachments/assets/2790294b-3846-4acf-9053-9cc01521808f" />

JAGADISHSUNILPEDNEKAR commented:
Hi @karask, just checking in on this PR! I’ve added PSBT (BIP-174) support with tests and examples. I’ve also mentioned some future enhancements in the PR that I’d love to work on next. Let me know your thoughts or if anything needs adjusting. Thanks!

karask commented:
Hi @JAGADISHSUNILPEDNEKAR !

Thank you for the contribution! I see you have done a lot of cleanup in addition to PSBT. I have not looked in detail the PSBT code or even tried to merge/run the code. It will take some time and I wanted to reply asap given the amount of effort you have put into this. For now I had a quick look at how the existing codebase is modified and check the new files that you have added. 

Some questions/comments:

- Why did you include all the imports in __init__.py?
- In block.py Why did you change the parsing of the block? (lines ~112-140)I find that the existing one (with header_format were the header structure is defined first is cleaner. Did you need to change the parsing for a specific reason?
- Are files apply_patches.py, new.py, combined_patch.py, combined_patch_final.py, combined_patch_v2.py, direct_patch.py, final_override.py, fix_bitcoin_utils.py, fix_tests.py, fix_transaction.py, hex_fix.py, main.py, monkey_patch.py, override_transaction.py, patch.runner.py, patch_functions.py, test_output_map.py, test_runner.py, tests/__init__.py, tests/test_from_raw.py, transaction_patch.py required?
- Why isn't bytes.fromhex sufficient in h_to_b in utils.py
- Are files mock_data.py, mock_test_data.py used to keep all the data in one place so as to clean the tests (which required cleaning :-) ? -- see next
- I also see heavy modifications of the existing tests. At first glance it seems that despite the mock_data changes above some test cases are missing..? (I'll check that when I try out the code).
- probably some instructions on how to add tests might be needed for future contributors

Again, thank you for the impressive PR. For future reference, I would suggest that you do several smaller PRs. It is quite challenging to review several changes at once. E.g. one PR to add PSBT, another PR to clean the tests with mock data, another PR for other minor cleanup like adding comments, etc.

Due to some time constraints I thought of skipping SoB this year. However, send me an email if you are interested to participate.

JAGADISHSUNILPEDNEKAR commented:
Hi @karask,

Thank you for taking the time to review my PR! I appreciate your feedback and quick response. Let me address your questions:

1. **Imports in __init__.py**: I added these imports to make the PSBT functionality readily available when users import the library. I can revert this and use a more minimal approach if you prefer.

2. **Changes in block.py**: The modifications in the block parsing logic were made to improve error handling and provide better compatibility with the new PSBT functionality. The header_format approach is indeed clean, but I found that the revised approach provided more consistent behavior when handling edge cases. I'm open to reverting these changes if you feel the original implementation is preferable.

3. **Multiple patch/fix files**: These files were added during development and testing to ensure compatibility across different environments. They're not essential for the core functionality and can be removed before merging. I should have cleaned them up before submitting the PR.My bad I should have been more careful and should have done this before submitting the PR as I was in a hurry I made this mistake.

4. **h_to_b in utils.py**: The enhanced version handles edge cases like whitespace, '0x' prefixes, and odd-length hex strings that bytes.fromhex() would reject. It also provides better error messages. This helps make the library more robust when dealing with user input.

5. **mock_data.py and mock_test_data.py**: Yes, exactly! These files centralize test data to make tests cleaner and more maintainable. They also help ensure consistent test coverage when making changes.

6. **Test modifications**: I've attempted to preserve all test cases while refactoring them to be more maintainable. If you've noticed missing cases, please point them out and I'll make sure they're included.

7. **Test documentation**: You're right - adding documentation on how to add tests would be valuable for future contributors. I can create this as part of this PR or as a separate follow-up.

I apologize for the size of this PR - I understand it makes reviewing more challenging. In the future, I'll submit smaller, more focused PRs as you suggested.

I'm happy to make any adjustments you recommend to ensure these changes integrate smoothly with the codebase. Would you like me to start by removing the unnecessary patch/fix files to make the PR cleaner?

Regarding Summer of Bitcoin, I'm participating in it . I was actually hoping that this project would be part of the program, and I'm excited by your mention of it. I'll send you an email with more details.

Thank you again for your feedback!

JAGADISHSUNILPEDNEKAR commented:
Subject: Interest in Summer of Bitcoin with python-bitcoin-utils

Respected Kostas Karasavvas Sir,

Thank you for considering my participation in Summer of Bitcoin despite
your time constraints. I'm very passionate about Bitcoin development, and
my work on the PSBT PR(#118) demonstrates my commitment to improving
python-bitcoin-utils.

I understand that time is limited, and I'd like to propose a mentorship
structure that would require minimal oversight:

1. We could set up weekly check-ins to discuss progress and direction
2. I could work on smaller, focused PRs that are easier to review
3. I'd document my work thoroughly, including test cases and examples
4. I could take on some maintenance tasks to reduce your workload

For the Summer of Bitcoin project, I could focus on:
- Completing the PSBT implementation with all refinements you suggested
- Improving test documentation and structure
- Adding support for additional Bitcoin features (e.g., Taproot, Lightning,
etc.)
- Creating comprehensive examples and tutorials for new users

I've already spent significant time understanding the codebase, which
should reduce the onboarding time needed. I'm also comfortable working
independently while following your guidance on the project's direction.

Please let me know if this approach would work for you or if you have other
suggestions that would make participation more feasible given your time
constraints.

Thank you for your consideration,
Jagadish Sunil Pednekar

On Thu, 13 Mar 2025 at 17:41, Kostas Karasavvas ***@***.***>
wrote:

> Hi @JAGADISHSUNILPEDNEKAR <https://github.com/JAGADISHSUNILPEDNEKAR> !
>
> Thank you for the contribution! I see you have done a lot of cleanup in
> addition to PSBT. I have not looked in detail the PSBT code or even tried
> to merge/run the code. It will take some time and I wanted to reply asap
> given the amount of effort you have put into this. For now I had a quick
> look at how the existing codebase is modified and check the new files that
> you have added.
>
> Some questions/comments:
>
>    - Why did you include all the imports in *init*.py?
>    - In block.py Why did you change the parsing of the block? (lines
>    ~112-140)I find that the existing one (with header_format were the header
>    structure is defined first is cleaner. Did you need to change the parsing
>    for a specific reason?
>    - Are files apply_patches.py, new.py, combined_patch.py,
>    combined_patch_final.py, combined_patch_v2.py, direct_patch.py,
>    final_override.py, fix_bitcoin_utils.py, fix_tests.py, fix_transaction.py,
>    hex_fix.py, main.py, monkey_patch.py, override_transaction.py,
>    patch.runner.py, patch_functions.py, test_output_map.py,
>    test_runner.py, tests/*init*.py, tests/test_from_raw.py,
>    transaction_patch.py required?
>    - Why isn't bytes.fromhex sufficient in h_to_b in utils.py
>    - Are files mock_data.py, mock_test_data.py used to keep all the data
>    in one place so as to clean the tests (which required cleaning :-) ? -- see
>    next
>    - I also see heavy modifications of the existing tests. At first
>    glance it seems that despite the mock_data changes above some test cases
>    are missing..? (I'll check that when I try out the code).
>    - probably some instructions on how to add tests might be needed for
>    future contributors
>
> Again, thank you for the impressive PR. For future reference, I would
> suggest that you do several smaller PRs. It is quite challenging to review
> several changes at once. E.g. one PR to add PSBT, another PR to clean the
> tests with mock data, another PR for other minor cleanup like adding
> comments, etc.
>
> Due to some time constraints I thought of skipping SoB this year. However,
> send me an email if you are interested to participate.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/karask/python-bitcoin-utils/pull/118#issuecomment-2721041460>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BLSDJOVVC6QSTHNZJ2TMTSL2UFYWDAVCNFSM6AAAAABYS3UFXGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRRGA2DCNBWGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: karask]*karask* left a comment (karask/python-bitcoin-utils#118)
> <https://github.com/karask/python-bitcoin-utils/pull/118#issuecomment-2721041460>
>
> Hi @JAGADISHSUNILPEDNEKAR <https://github.com/JAGADISHSUNILPEDNEKAR> !
>
> Thank you for the contribution! I see you have done a lot of cleanup in
> addition to PSBT. I have not looked in detail the PSBT code or even tried
> to merge/run the code. It will take some time and I wanted to reply asap
> given the amount of effort you have put into this. For now I had a quick
> look at how the existing codebase is modified and check the new files that
> you have added.
>
> Some questions/comments:
>
>    - Why did you include all the imports in *init*.py?
>    - In block.py Why did you change the parsing of the block? (lines
>    ~112-140)I find that the existing one (with header_format were the header
>    structure is defined first is cleaner. Did you need to change the parsing
>    for a specific reason?
>    - Are files apply_patches.py, new.py, combined_patch.py,
>    combined_patch_final.py, combined_patch_v2.py, direct_patch.py,
>    final_override.py, fix_bitcoin_utils.py, fix_tests.py, fix_transaction.py,
>    hex_fix.py, main.py, monkey_patch.py, override_transaction.py,
>    patch.runner.py, patch_functions.py, test_output_map.py,
>    test_runner.py, tests/*init*.py, tests/test_from_raw.py,
>    transaction_patch.py required?
>    - Why isn't bytes.fromhex sufficient in h_to_b in utils.py
>    - Are files mock_data.py, mock_test_data.py used to keep all the data
>    in one place so as to clean the tests (which required cleaning :-) ? -- see
>    next
>    - I also see heavy modifications of the existing tests. At first
>    glance it seems that despite the mock_data changes above some test cases
>    are missing..? (I'll check that when I try out the code).
>    - probably some instructions on how to add tests might be needed for
>    future contributors
>
> Again, thank you for the impressive PR. For future reference, I would
> suggest that you do several smaller PRs. It is quite challenging to review
> several changes at once. E.g. one PR to add PSBT, another PR to clean the
> tests with mock data, another PR for other minor cleanup like adding
> comments, etc.
>
> Due to some time constraints I thought of skipping SoB this year. However,
> send me an email if you are interested to participate.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/karask/python-bitcoin-utils/pull/118#issuecomment-2721041460>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BLSDJOVVC6QSTHNZJ2TMTSL2UFYWDAVCNFSM6AAAAABYS3UFXGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRRGA2DCNBWGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

karask commented:
Hi @JAGADISHSUNILPEDNEKAR ,

Please also add the tests added by PR #120  in the new framework!

JAGADISHSUNILPEDNEKAR commented:
Hi @karask,

Thank you for your feedback. I've addressed all the requested changes:

1. Cleaned up `__init__.py` by removing the PSBT-specific imports
2. Reverted the changes to the block parsing code in `block.py` to maintain the original header_format approach
3. Removed all the unnecessary patch/fix files that were added during development
4. Preserved the original `h_to_b` implementation in `utils.py` as a comment with an explanation of the edge cases handled by the new version
5. Added a `TESTS_README.md` file explaining the new testing framework and how to add tests

Regarding point #6 (incorporating tests from PR #120), I plan to address this in my next update. I wanted to get these initial changes to you first for review.

Please let me know if there are any other adjustments needed.

Thank you!
